### PR TITLE
Use lax.scan inside jnp.polyval

### LINKED ISSUE #3047: polyval is exceedingly slow to compile for high degree polynomials
In my latest attempt at Twitter Driven Development (TM), I wanted to see if JAX could be [as fast as Julia](https://twitter.com/sethaxen/status/1259973355869925376) for evaluating a 10,000 degree polynomials.

Unfortunately, the implementation of `polyval` uses a Python `for` loop, so I can't even run this example because it takes forever to run and then crashes:
```python
import jax.numpy as jnp
import jax
import numpy as np

x = np.random.rand()
p = np.random.randn(10000)
jnp.polyval(p, x)  # works
jax.jit(jnp.polyval)(p, x)  # never finishes
```

There are probably more sophisticated algorithms (like whatever @sethaxen was using here) but a simple improvement would be to switch from a Python `for` loop to `lax.scan`.

### PR DESCRIPTION
**Pull request by shoyer**
Fixes https://github.com/google/jax/issues/3047

Scan solves the slow compilation problem.

I've also removed any attempt at support for numpy.poly1d objects. I doubt @mattjj ever really intended to support them in 9c49a9bfe690738d502a48e33d2f421c4694e6ab and the example from the np.polyval docstring already crashes on master:
```
TypeError: Argument '   2
3 x + 1' of type <class 'numpy.poly1d'> is not a valid JAX type
```

### CODE REVIEW

File: jax/numpy/lax_numpy.py
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:2549]: I suspect the downside here is that this will be slower on GPU because loops are often slow on GPU. Perhaps that's an acceptable tradeoff? Or perhaps there is something to fixing lax.scan to unroll a little bit internally on GPU?

Comment (mattjj) [reply jax/numpy/lax_numpy.py:2549]: Unrolling a bit on GPU sounds reasonable to me! (Though I wonder if that's also an easy XLA:GPU-level change, for cases when the While HLO in question is really a "for loop with static trip count", as is already detected in some optimizations.)

Comment (shoyer) [reply jax/numpy/lax_numpy.py:2549]: I thought XLA:GPU could handle static loop sizes well, but ouch, it is slow!

256 ms to do a single iteration on GPU, vs 31 us on CPU!

**Comment by shoyer** [comment]
I added partial unrolling, though this really seems like something that should be pushed into `lax.scan` or `lax.while_loop`.

**Review by hawkinsp — APPROVED** [review]
I agree this logic should live in `scan` or in XLA perhaps. Fine for now.

**Comment by hawkinsp** [comment]
It's probably worth digging into why XLA's static trip count detection logic didn't fire here.

**Comment by mattjj** [comment]
> It's probably worth digging into why XLA's static trip count detection logic didn't fire here.

My guess was it did fire, but it doesn't lead to unrolling (in the sense that there are still separate kernels dispatched for each iteration of the loop body, though only a fixed number of such kernels are dispatched and we don't have to pull a value back to the host to decide when to terminate the loop). Is that different from your understanding?

File: jax/numpy/lax_numpy.py
Comment (shoyer) [inline jax/numpy/lax_numpy.py:2566]: This use of `reshape` has a bug: it won't work for higher dimensional `p`.

**Comment by hawkinsp** [comment]
Is this blocked on anything? I just want to make sure we haven't dropped the ball.

**Comment by shoyer** [comment]
I think my loop unrolling has a bug which isn't triggered by any of the unit tests currently. That needs to be fixed before submitting.

Could also be nice to try to use @froystig's https://github.com/google/jax/pull/3159 here.

File: jax/numpy/lax_numpy.py
Comment (shoyer) [inline jax/numpy/lax_numpy.py:2726]: This `unroll` number really seems to demand platform specific tuning.

On my laptop's CPU, `unroll=8` is about 30% faster than `unroll=1`, whereas `unroll=64` is 4x slower.

I haven't tested it on GPUs yet but given the cost of kernel launches suspect the conclusions would be pretty different.

**Comment by jekbradbury** [comment]
Scan now supports unrolling natively. Any chance you could revive this, @shoyer? A user just pinged the issue.

**Comment by shoyer** [comment]
So I actually think the current version of this PR (which uses scan's builtin unrolling) is basically ready to merge. The only remaining question is how to pick the unroll parameter.

I did some quick tests to benchmark compile time and runtime CPU/GPU/TPU as a function of the `unroll` parameter, for all powers of two between 1 and 128: https://gist.github.com/shoyer/4e0328c277e46f58c47d79b85a51aa0a

It's a hard to find a clear winner:
- For CPU, `unroll=8` or `unroll=16` seems optimal. Higher values of `unroll` are weirdly 4-10x slower.
- For GPU, maybe `unroll=128` or even higher, if slow compilation is not problematic.
- For TPU, maybe `unroll=64`?

`unroll=16` seems pretty safe across the board, though maybe we should also make this a custom JAX-specific parameter that users can adjust?

**Comment by shoyer** [comment]
> `unroll=16` seems pretty safe across the board, though maybe we should also make this a custom JAX-specific parameter that users can adjust?

OK, I just pushed a revision that does exactly this. PTAL

**Comment by jekbradbury** [comment]
Seems reasonable to me!

**Comment by shoyer** [comment]
@jekbradbury want to do a final review here?

**Review by jekbradbury — APPROVED** [review]

File: jax/_src/numpy/lax_numpy.py
Comment (soraros) [inline jax/_src/numpy/lax_numpy.py:4042]: Nit: `functools.partial` -> `partial`

Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:4042]: Thanks! Fixed in #7776

**Pull request merged**
